### ISSUE TITLE
RememberMe (auth token) cookies not using domain from 'session.cookie.domain'

### ISSUE THREAD

**Issue by renklaf** [simplesamlphp/simplesamlphp#770]
Currently when setting the auth token cookie, the config 'session.cookie.domain' isn't used. 

When running SimpleSAMLphp in proxy mode with hosted IdP's on the same domain, the proxy will not be able to resolve the auth token cookie set by one of the hosted IdP's on a different subdomain. 

I have created a patch for this in the method below, that I can create a pull request for if the issue is accepted. 

(Original code)
https://github.com/simplesamlphp/simplesamlphp/blob/a5437da2dbbb8dc6348cac02dc343c03bc7f3e14/lib/SimpleSAML/Session.php#L548-L560

**Comment by jaimeperez** [comment]
Hi @renklaf!

I'm not sure I'm following. The _remember me_ mechanism has nothing to do with the _auth token_ cookie. The former allows the IdP to keep track of the user so that the username is pre-filled, while the latter is an anti-session-fixation mechanism.

Also, I don't see why a SimpleSAMLphp instance running as a proxy should honor the cookies set by other instances running as IdPs.

Would you mind clarifying a bit what you are trying to do?

**Comment by renklaf** [comment]
The proposed patch

```diff
diff --git a/unpatched_simplesamlphp-1.15/lib/SimpleSAML/Session.php b/patched_simplesamlphp-1.15/lib/SimpleSAML/Session.php
index f7bb281..bf08178 100755
--- a/unpatched_simplesamlphp-1.15/lib/SimpleSAML/Session.php
+++ b/patched_simplesamlphp-1.15/lib/SimpleSAML/Session.php
@@ -760,11 +760,17 @@ class SimpleSAML_Session implements Serializable
         }

         if ($this->authToken !== null) {
+
+            $cookieParams = $sessionHandler->getCookieParams();
+            if ($params !== null) {
+                $cookieParams = array_merge($cookieParams, $params);
+            }
+
             $globalConfig = SimpleSAML_Configuration::getInstance();
             \SimpleSAML\Utils\HTTP::setCookie(
                 $globalConfig->getString('session.authtoken.cookiename', 'SimpleSAMLAuthToken'),
                 $this->authToken,
-                $params
+                $cookieParams
             );
         }
     }
```

**Comment by jaimeperez** [comment]
Hi again @renklaf!

Thanks a lot for the detailed explanation. Now I see what you meant.

To be honest, I've always thought the _remember me_ mechanism was flawed from the beginning, and should have never been included in SimpleSAMLphp as is. It lacks proper documentation and it's not really usable in large-scale deployments.

Would you mind opening a PR with the fix? Also, I think we could easily reduce it to a one-liner:

```php
    /**
     * Update session cookies.
     *
     * @param array $params The parameters for the cookies.
     */
    public function updateSessionCookies($params = null)
    {
        $sessionHandler = \SimpleSAML\SessionHandler::getSessionHandler();

        $params = $params ?: array_merge($sessionHandler->getCookieParams(), $params);

        if ($this->sessionId !== null) {
            $sessionHandler->setCookie($sessionHandler->getSessionCookieName(), $this->sessionId, $params);
        }

        if ($this->authToken !== null) {
            $globalConfig = SimpleSAML_Configuration::getInstance();
            \SimpleSAML\Utils\HTTP::setCookie(
                $globalConfig->getString('session.authtoken.cookiename', 'SimpleSAMLAuthToken'),
                $this->authToken,
                $params
            );
        }
```

Note that proper cookie parameters should be used for both the SimpleSAMLphp session cookie **and** the auth token. The bug affects both, as both are updated by `updateSessionCookies()`, and both use the same `$params` parameter.

**Comment by renklaf** [comment]
Hi @jaimeperez 

Thanks! I'll open a PR today. I noticed that `SessionHandler->setCookie(..) `calls `getCookieParams` internally.

Could a simpler solution be the following?
```php
    /**
     * Update session cookies.
     *
     * @param array $params The parameters for the cookies.
     */
    public function updateSessionCookies($params = null)
    {
        $sessionHandler = \SimpleSAML\SessionHandler::getSessionHandler();

        if ($this->sessionId !== null) {
            $sessionHandler->setCookie($sessionHandler->getSessionCookieName(), $this->sessionId, $params);
        }

        if ($this->authToken !== null) {
            $globalConfig = SimpleSAML_Configuration::getInstance();
            $sessionHandler->setCookie(
                $globalConfig->getString('session.authtoken.cookiename', 'SimpleSAMLAuthToken'), 
                $this->authToken, 
                $params
            );
        }
    }

**Comment by jaimeperez** [comment]
You are absolutely right, I didn't notice that that `setCookie()` method was coming from different classes when setting each cookie. That should work, I think.

**Comment by jaimeperez** [comment]
Fixed in #772.
